File: app/src/main/java/igrek/todotree/service/tree/persistence/TreePersistenceService.java
package igrek.todotree.service.tree.persistence;


import igrek.todotree.domain.treeitem.AbstractTreeItem;
import igrek.todotree.exceptions.DeserializationFailedException;

public class TreePersistenceService {
	
	private GsonTreeDeserializer deserializer = new GsonTreeDeserializer();
	private JsonTreeSerializer serializer = new JsonTreeSerializer();
	
	public TreePersistenceService() {
	}
	
	public String serializeTree(AbstractTreeItem root) {
		return serializer.serializeTree(root);
	}
	
	public AbstractTreeItem deserializeTree(String data) throws DeserializationFailedException {
		return deserializer.deserializeTree(data);
	}
	
}

File: app/src/main/java/igrek/todotree/service/tree/persistence/TreePersistenceService.kt
package igrek.todotree.service.tree.persistence

import igrek.todotree.domain.treeitem.AbstractTreeItem
import igrek.todotree.exceptions.DeserializationFailedException

class TreePersistenceService {

    private val serializer = YamlTreeSerializer()
    private val deserializer = YamlTreeDeserializer()
    private val fallbackDeserializer = JsonKtxTreeDeserializer()

    fun serializeTree(root: AbstractTreeItem): String {
        return serializer.serializeTree(root)
    }

    @Throws(DeserializationFailedException::class)
    fun deserializeTree(data: String): AbstractTreeItem {
        return try {
            deserializer.deserializeTree(data)
        } catch (e: DeserializationFailedException) {
            fallbackDeserializer.deserializeTree(data)
        }
    }
}